CSP violations are reported in the console twice in case preload scanner kicks in

youennf opened the pull request:
#### 58bccea8c81b84c2bf53b799e47a996c114a7745
<pre>
CSP violations are reported in the console twice in case preload scanner kicks in
<a href="https://rdar.apple.com/149210434">rdar://149210434</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291524">https://bugs.webkit.org/show_bug.cgi?id=291524</a>

Reviewed by Alex Christensen.

A subresource load may be started as a preload via the preload scanner and quickly after as a regular subresource load.
Before the patch, CSP would report the violation twice, once for the preload scanner and once for the regular load.
We disable report violation for the preload scanner as it is redundant information and may trigger flakinesses in tests like in <a href="https://bugs.webkit.org/show_bug.cgi?id=291436.">https://bugs.webkit.org/show_bug.cgi?id=291436.</a>

* LayoutTests/http/tests/loading/do-not-preload-css-blocked-by-csp-expected.txt:
* LayoutTests/http/tests/loading/do-not-preload-script-src-blocked-by-csp-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/wasm/webapi/esm-integration/script-src-blocks-wasm.tentative.sub-expected.txt:
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestImage):
(WebCore::CachedResourceLoader::allowedByContentSecurityPolicy const):
(WebCore::CachedResourceLoader::canRequest):
(WebCore::CachedResourceLoader::requestResource):
* Source/WebCore/loader/cache/CachedResourceLoader.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::logToConsole const):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:

Canonical link: <a href="https://commits.webkit.org/293751@main">https://commits.webkit.org/293751@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/402ce06b5067c095eb5ad02a2de0355d65022bd1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99838 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19483 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9766 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104964 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50416 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101879 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19788 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27919 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75995 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33084 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102845 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/15083 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90155 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56360 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14888 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8142 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49788 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84807 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8227 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107324 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26949 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19679 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84949 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27311 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86360 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84475 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21444 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29151 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6871 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20758 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26885 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32096 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26696 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30013 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28257 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/csp/ContentSecurityPolicy.h:
```diff
@@ -291,6 +293,7 @@ class ContentSecurityPolicy final : public CanMakeThreadSafeCheckedPtr<ContentSe
     SandboxFlags m_sandboxFlags;
     bool m_overrideInlineStyleAllowed { false };
     bool m_isReportingEnabled { true };
+    bool m_isReportingToConsoleEnabled { true };
```
Would it be worthwhile to make all the bools together and make them bit fields?

youennf replied:
OK

achristensen07 commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -561,6 +562,11 @@ bool CachedResourceLoader::allowedByContentSecurityPolicy(CachedResource::Type t
 
+    contentSecurityPolicy->setIsReportingToConsoleEnabled(canReportViolationAsConsoleMessage);
+    auto scope = makeScopeExit([&] {
+        contentSecurityPolicy->setIsReportingToConsoleEnabled(true);
```
Shouldn't we restore the old value, whatever it was?  This assumes the old value was false.

youennf replied:
It is just a way to temporarily disable console reporting, so we should always set it back to true.

achristensen07 commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -657,7 +663,8 @@ bool CachedResourceLoader::canRequest(CachedResource::Type type, const URL& url,
         }
 
-        if (!allowedByContentSecurityPolicy(type, url, options, ContentSecurityPolicy::RedirectResponseReceived::No))
+        bool canReportViolationAsConsoleMessage = forPreload == ForPreload::No || isLinkPreload;
```
nit: I'd say "should" instead of "can" throughout this PR.

youennf replied:
OK

youennf committed:
CSP violations are reported in the console twice in case preload scanner kicks in
rdar://149210434
https://bugs.webkit.org/show_bug.cgi?id=291524

Reviewed by Alex Christensen.

A subresource load may be started as a preload via the preload scanner and quickly after as a regular subresource load.
Before the patch, CSP would report the violation twice, once for the preload scanner and once for the regular load.
We disable report violation for the preload scanner as it is redundant information and may trigger flakinesses in tests like in https://bugs.webkit.org/show_bug.cgi?id=291436.

* LayoutTests/http/tests/loading/do-not-preload-css-blocked-by-csp-expected.txt:
* LayoutTests/http/tests/loading/do-not-preload-script-src-blocked-by-csp-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/wasm/webapi/esm-integration/script-src-blocks-wasm.tentative.sub-expected.txt:
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestImage):
(WebCore::CachedResourceLoader::allowedByContentSecurityPolicy const):
(WebCore::CachedResourceLoader::canRequest):
(WebCore::CachedResourceLoader::requestResource):
* Source/WebCore/loader/cache/CachedResourceLoader.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::logToConsole const):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:

Canonical link: https://commits.webkit.org/293751@main
